hcipy/optics/segmented_mirror.py

```diff
@@ -0,0 +1,237 @@
+        -------
+        Wavefront
+            The reflected wavefront.
+        """
+        wf = wavefront.copy()
+        wf.electric_field *= np.exp(2j * self.surface * wavefront.wavenumber)
+        return wf
+
+    def backward(self, wavefront):
+        """Propagate a wavefront backwards through the deformable mirror.
+
+        Parameters
+        ----------
+        wavefront : Wavefront
+            The incoming wavefront.
+
+        Returns
+        -------
+        Wavefront
+            The reflected wavefront.
+        """
+        wf = wavefront.copy()
+        wf.electric_field *= np.exp(-2j * self.surface * wavefront.wavenumber)
+        return wf
+
+    @property
+    def surface(self):
+        """ The surface of the segmented mirror in meters, the full surface as a Field.
+        """
+        surf = self.apply_coef()
+        return surf
+
+    @property
+    def coef(self):
+        """ The surface shape of the deformable mirror, in meters and radians; PTT segment coefficients.
+        """
+        return self._coef
+
+    def flatten(self):
+        """ Flatten the DM by setting all segment coefficients to zero."""
+        self._coef[:] = 0
+
+    def set_segment(self, segid, piston, tip, tilt):
+        """ Set an individual segment of the DM.
+
+        Piston in meter of surface, tip and tilt in radians of surface.
+
+        Parameters
+        -------------
+        segid : integer
+            Index of the segment you wish to control, starting at 1 (center would  be 0, but doesn't exist)
+        piston, tip, tilt : floats, meters and radians
+            Piston (in meters) and tip and tilt (in radians)
+        """
+        self._coef[segid - 1] = [piston, tip, tilt]
+
+    def _setup_grids(self):
+        """ Set up the grids to compute the segmented mirror surface into.
+        This is relatively slow, but we only need to do this once for
+        each size of input grids.
+        """
+        npix = self.aperture.shaped.shape[0]
+        if npix == self._last_npix:
+            return
+        else:
+            self._last_npix = npix
+
+        x, y = self.input_grid.coords
+
+        self._seg_mask = np.zeros_like(x)
+        self._seg_x = np.zeros_like(x)
+        self._seg_y = np.zeros_like(y)
+        self._seg_indices = dict()
+
+        pupil_grid = make_pupil_grid(dims=npix, diameter=PUP_DIAMETER)
+        aper_num, seg_positions = get_atlast_aperture(normalized=False,
+                                                      segment_transmissions=np.arange(1, self.segnum + 1))
+        aper_num = evaluate_supersampled(aper_num, pupil_grid, 2)
+
+        self._seg_mask = np.copy(aper_num)
+
+        for i in self.segmentlist:
+            wseg = np.where(self._seg_mask == i)
+            self._seg_indices[i] = wseg
+
+            cenx, ceny = self.seg_pos.points[i - 1]
+
+            self._seg_x[wseg] = x[wseg] - cenx
+            self._seg_y[wseg] = y[wseg] - ceny
+
+            # Set gaps to zero
```

Comment:
When I initially tested this, the gaps between the segments had super high values that would mess with the actual phase ramps on the segments. I solved it by defining a threshold and setting anything above it to zero, but I feel like there might be a more robust solution to this.

Reply:
The problem here is that for each pixel it may not be perfectly clear to which segment it belongs. This should probably be changed in the `make_segmented_aperture()` function. Maybe this function should be able to return the `subaperture_indices`?

Reply:
See https://github.com/ehpor/hcipy/pull/15#issuecomment-495401660